specs/merkle-tree.md

```diff
@@ -0,0 +1,158 @@
+
+So for example, if we intended to run some set of transactions on Goerli (5), OP Goerli (420), and Polygon Mumbai (80001). Then the leaves are expected to be ordered with the approval leaf for Goerli first, then all of the transaction leaves on Goerli in ascending order, then the approval leaf for OP Goerli, all the transaction leaves on OP Goerli in ascending order, and so on.
+
+### 3. Every chain ID and index combination must be unique within the tree
+Since the Sphinx Merkle tree leaves contain explicit indexes, it is also possible to construct trees that contain multiple leaves with the same index value and chain ID. However, only one of the leaves would be executable on-chain. Therefore, constructing a tree with multiple leaves with the same chain ID and index would make it ambiguous to the relayer which leaf should be executed, and would result in a tree that cannot be fully executed. So we impose the standard that in a given Sphinx Merkle tree there must be at most one leaf with each chain ID and index combination.
+
+### 4. There must be exactly one approval leaf or cancel leaf per chain
+Since Sphinx Merkle trees can contain an arbitrary set of leaves, it is possible to construct a single tree that contains `APPROVE` or `CANCEL` leaves that approve or cancel multiple separate deployments on a single chain. However, this leads to an ambiguous situation for the relayer where it is unclear which deployment should be approved or canceled.
+
+Additionally, The `SphinxModuleProxy` does not support both canceling an existing deployment and approving a new deployment in the same tree, therefore any tree that contains both a `CANCEL` and `APPROVE` leaf for a given chain would not be executable on that chain. So we impose the standard, that in a given Sphinx Merkle tree there should be exactly one `APPROVE` or `CANCEL` leaf per chain.
+
+If the `arbitraryChain` field is set to true, then we consider this approval leaf to apply to all chains. Therefore if `arbitraryChain` = true, then there should be exactly one `APPROVE` leaf in the entire tree and no `CANCEL` leaves.
+
+## Sphinx Merkle Tree Generation Standard Implementation
+We provide a standard implementation for generating Sphinx Merkle trees that adhere to the above invariants. This implementation is used by the official Sphinx Foundry plugin.
+
+### Input
+Accepts an object where the keys are canonical chain ids and the values are deployment data objects which contain all of the necessary info to assemble a SphinxMerkleTree. See the [inline documentation](TODO(end)) for exact details on the input data. The Sphinx Merkle tree generation function should be agnostic to the source of the transaction data so it can be used to assemble a tree based on transactions generated from any scripting framework (i.e Foundry, Hardhat Ignition, some arbitrary future framework).
```

Comment:
> Accepts an object where the keys are canonical chain ids and the values are deployment data objects which contain all of the necessary info to assemble a SphinxMerkleTree.

I think we should explicitly mention and link to the `DeploymentData` object. Something like "Accepts a [`DeploymentData`](TODO(end)) object, which maps canonical chain IDs to..."